cf/mixin/coordinate.py

```diff
@@ -390,18 +412,220 @@ def autoperiod(self, inplace=False):
+
+        * The "standard_name" property.
+        * The "id" attribute, preceded by ``'id%'``.
+        * The "cf_role" property, preceded by ``'cf_role='``.
+        * The "axis" property, preceded by ``'axis='``.
+        * The "long_name" property, preceded by ``'long_name='``.
+        * All other properties (including "standard_name"), preceded by
+          the property name and an ``'='``.
+        * The coordinate type (``'X'``, ``'Y'``, ``'Z'`` or ``'T'``).
+        * The netCDF variable name, preceded by ``'ncvar%'``.
+
+        .. versionadded:: 3.0.0
+
+        .. seealso:: `id`, `identity`
+
+        :Parameters:
+
+            {{generator: `bool`, optional}}
+
+            ctypes: (sequence of) `str`
+                If set then return the coordinate type (if any) as the
+                first identity and restrict the possible coordinate
+                types to be any of these characters. By default, a
+                coordinate type is the last identity. Setting to a
+                subset of ``'XTYZ'`` can give performance
+                improvements, as it will reduce the number of
+                coordinate types that are checked in circumstances
+                when particular coordinate types have been ruled out a
+                priori. If a coordinate type is omitted then it will
+                not be in the returned identities even if the
+                coordinate construct is of that type. Coordinate types
+                are checked in the order given.
+
+                *Parameter example:*
+                  ``ctypes='Y'``
+
+                *Parameter example:*
+                  ``ctypes='XY'``
+
+                *Parameter example:*
+                  ``ctypes=('T', 'X')``
+
+        :Returns:
+
+            `list`
+                The identities.
+
+        **Examples:**
+
+        >>> f.properties()
+        {'foo': 'bar',
+         'long_name': 'Air Temperature',
+         'standard_name': 'air_temperature'}
+        >>> f.nc_get_variable()
+        'tas'
+        >>> f.identities()
+        ['air_temperature',
+         'long_name=Air Temperature',
+         'foo=bar',
+         'standard_name=air_temperature',
+         'ncvar%tas']
+
+        """
+        if ctypes:
+            pre = (self._ctypes_iter(ctypes),)
+            pre0 = kwargs.pop("pre", None)
+            if pre0:
+                pre = tuple(pre0) + pre
+
+            kwargs["pre"] = pre
+        else:
+            post = (self._ctypes_iter("XTYZ"),)
+            post0 = kwargs.pop("post", None)
+            if post0:
+                post += tuple(post0)
+
+            kwargs["post"] = post
+
+        return super().identities(generator=generator, **kwargs)
+
+    def _ctypes_iter(self, ctypes):
+        """Generator for returning the coordinate type letter."""
+        for c in ctypes:
+            if getattr(self, c):
+                # This coordinate construct is of this type
+                yield c
+                return
```

Comment:
Is the `return` at the right indentation level here? Just to check since it logically I would have thought it should go one level down.

Reply:
I think this is OK, as we want this method to short circuit on the first coordinate type (T, X, Y or z) that gives a match.

Reply:
Ah, yes, I've played around with a toy example of an equivalent structure and I think I see now. :+1: Resolving this.